src/ServiceControl/Monitoring/HeartbeatMonitor.cs

```diff
@@ -0,0 +1,53 @@
+namespace ServiceControl.Monitoring
+{
+    using System;
+    using System.Threading;
+
+    class HeartbeatMonitor
+    {
+        volatile RecordedHeartbeat heartbeat = new RecordedHeartbeat(HeartbeatStatus.Unknown, null);
+
+        public void MarkAlive(DateTime timestamp)
+        {
+            var newReading = new RecordedHeartbeat(HeartbeatStatus.Alive, timestamp);
+            var done = false;
+
+            while (!done)
+            {
+                var priorReading = heartbeat;
+
+                if (priorReading.Timestamp.GetValueOrDefault() <= timestamp)
+                {
+                    done = Interlocked.CompareExchange(ref heartbeat, newReading, priorReading) == priorReading;
+                }
+                else
+                {
+                    done = true;
+                }
+            }
+        }
+
+        public RecordedHeartbeat Age(DateTime cutoff)
+        {
+            var done = false;
+
+            while (!done)
+            {
+                var priorReading = heartbeat;
+                if (priorReading.Timestamp.GetValueOrDefault() < cutoff)
+                {
+                    var newReading = new RecordedHeartbeat(HeartbeatStatus.Dead, priorReading.Timestamp);
+                    done = Interlocked.CompareExchange(ref heartbeat, newReading, priorReading) == priorReading;
```

Comment:
We can get around all of that by just locking the heartbeat when we update it but it means a lot of locking and synchronization, especially when SC first starts and has a backlog of heartbeats it's trying to process concurrently.